src/workerd/api/global-scope.c++

```diff
@@ -386,21 +387,42 @@ kj::Promise<WorkerInterface::AlarmResult> ServiceWorkerGlobalScope::runAlarm(
 
     auto& alarm = KJ_ASSERT_NONNULL(handler.alarm);
 
-    auto alarmResultPromise = context
-        .run([exportedHandler, &alarm,
+    return context
+        .run([exportedHandler, &context, timeoutMs = timeout, &alarm,
               maybeAsyncContext = jsg::AsyncContextFrame::currentRef(lock)]
              (Worker::Lock& lock) mutable -> kj::Promise<WorkerInterface::AlarmResult> {
       jsg::AsyncContextFrame::Scope asyncScope(lock, maybeAsyncContext);
+
+      // We want to limit alarm handler walltime to 15 minutes at most. If the timeout promise
+      // completes we want to cancel the alarm handler. If the alarm handler promise completes first
+      // timeout will be canceled.
+      auto timeoutPromise = context.afterLimitTimeout(timeoutMs*kj::MILLISECONDS).then([&context]() -> kj::Promise<WorkerInterface::AlarmResult> {
+        LOG_NOSENTRY(WARNING, "Alarm exceeded its allowed execution time");
+        // We don't want to delete the alarm sicne we have not successfully completed the alarm
+        // execution.
+        auto& actor = KJ_ASSERT_NONNULL(context.getActor());
+        auto& persistent = KJ_ASSERT_NONNULL(actor.getPersistent());
+        persistent.cancelDeferredAlarmDeletion();
+        // We don't want the handler to keep running after timeout.
+        context.abort(JSG_KJ_EXCEPTION(FAILED, Error, "Alarm exceeded its allowed execution time"));
+        // For now, we don't want to make the timeout destructive. As such, we will mark it as
+        // retriable, but let's now count retries against the limit. This behaviour will most likely
+        // change in the near future.
+        // TODO(soon): Make alarm timeouts not retriable.
+        return WorkerInterface::AlarmResult {
+          .retry = true,
+          .retryCountsAgainstLimit = false,
+          .outcome = EventOutcome::EXCEPTION
```

Comment:
FWIW, it looks like the status for exceeding the `scheduledTimeoutMs` is `EventOutcome::EXCEEDED_CPU`?

Reply:
I'm not sure that we want to treat this error as `EventOutcome::EXCEEDED_CPU`. Ideally we'd have a specific reason for this, but since cron uses `EXCEEDED_CPU` I guess we can go with that for now.